feat: add `debug_getHistoricalSummaries` endpoint

acolytec3 opened the pull request:
**Motivation**

The Portal Network needs a way to access the `HistoricalSummaries` property on the `BeaconState` for post Capella beacon state since the HistoricalSummaries are used to anchor EL block header proofs to the beacon state roots in light client updates that the serve as the source of truth for verifying the authenticity of post capella EL block headers.  This currently requires downloading the entire Beacon State from a beacon node willing to serve the `debug` endpoints and then throwing away everything except for the desired data (or else downloading an ERA file from a willing endpoint).  It is desirable to be able to access this data more efficiently directly from beacon nodes.  

**Description**

This adds a new `getHistoricalSummaries` endpoint to the `debug` namespace on the REST API that serves the `HistoricalSummaries` field from the current `head` BeaconState as SSZ bytes.  We only need the most current one for use in constructing proofs of recent EL block headers embedded in a beacon block.

This mirrors similar [experimental work being done in the nimbus client](https://github.com/status-im/nimbus-eth2/pull/6675)

**Steps to test or reproduce**

```sh
git checkout historicalSummaries-endpoint
./lodestar beacon \
--rest.namespace='*' \
--checkpointSyncUrl=https://beaconstate.info \
```

In a separate terminal
```sh
curl -X GET --header "Accept: application/octet-stream" "http://127.0.0.1:9596/eth/v1/debug/historical_summaries" > ./hs.ssz
```

Then, use a script like below to verify that historical summaries are correctly returned
```ts
    import { ssz } from '@lodestar/types'

    const summariesBytes = fs.readFileSync('./test/util/hs.ssz')
    const summaries = ssz.capella.BeaconState.fields.historicalSummaries.deserialize(summariesBytes)
    console.log(summaries)
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ChainSafe/lodestar?pullRequest=7245) <br/>All committers have signed the CLA.

acolytec3 committed:
feat: add new getHistoricalSummaries endpoint to debug namespace

nflaig commented on packages/api/src/beacon/routes/debug.ts:
```diff
@@ -196,5 +203,15 @@ export function getDefinitions(_config: ChainForkConfig): RouteDefinitions<Endpo
       },
     },
+    getHistoricalSummaries: {
+      url: "/eth/v1/debug/historical_summaries",
```
If this endpoint is useful for Portal it makes sense to standardize it on the beacon-api or at least get rough consensus there

if we wanna add a non-standardized api we should rather go with v0 or add to lodestar namespace instead of debug

acolytec3 replied:
Fair point and this is early days still.  I looked at the Nimbus PR again and it looks like they moved it under a client specific namespace so maybe I should mirror this on the lodestar side for now?

nflaig replied:
Yeah, it would be better to move the api there but I am also fine with v0, although it's not clear if this should be in debug or beacon namespace

acolytec3 replied:
Understood.  I've switched to v0 for the moment.  I chatted @deme who's working on the Nimbus equivalent and he's hoping to get a PR for the consensus specs to further general conversation there sometime soon.

nflaig commented on packages/api/src/beacon/routes/debug.ts:
```diff
@@ -196,5 +203,15 @@ export function getDefinitions(_config: ChainForkConfig): RouteDefinitions<Endpo
+      resp: {
+        data: ssz.capella.HistoricalSummaries,
+        meta: EmptyMetaCodec,
+        onlySupport: WireFormat.ssz,
```
any reason why this shouldn't also support JSON?

acolytec3 replied:
No specific reason other than it's most useful for portal clients to download the ssz encoded variant as it requires less work to integrate it into our own work.  Will add a JSON option as well.

nflaig replied:
Generally we support both return formats and the client decides based on Accept header, if header is not present we default to JSON

acolytec3 replied:
This has been updated in the latest commit to support both.

acolytec3 committed:
Add JSON response

g11tech reviewed (commented):
lgtm, we can always move the endpoint once we have it in beacon apis, CI running

nflaig commented on packages/beacon-node/src/api/impl/debug/index.ts:
```diff
@@ -85,5 +86,21 @@ export function getDebugApi({
       };
     },
+    async getHistoricalSummaries(_, context) {
+      const {state} = await getStateResponseWithRegen(chain, "head");
```
imo hard coding this to `head` is not great, we should probably align this with other apis which take a `state_id` parameter

g11tech replied:
correct, stateid in the route and corresponding change here makes sense

acolytec3 replied:
Will do.  I also need to add an additional object to the response, a single proof that anchors the `historicalSummaries` property in the `BeaconState` and that will take me a bit to correctly figure out.  I have the code for the proof already setup on the ultralight side but need to match it up here.

acolytec3 replied:
Have updated here as well, adding the proof that anchors the `HistoricalSummaries` object in the `BeaconState`.

acolytec3 committed:
Restructure to use stateId and add proof to response

acolytec3 committed:
add test scaffolding

nflaig commented on packages/api/test/unit/beacon/genericServerTest/debug.test.ts:
```diff
@@ -58,5 +58,21 @@ describe("beacon / debug", () => {
       expect(res.wireFormat()).toBe(WireFormat.ssz);
       expect(toHexString(res.ssz())).toBe(toHexString(stateSerialized));
     });
+    it("getHistoricalSummaries", async () => {
```
This doesn't really test a lot which would not be already covered by our generic server tests.

To make the generic tests work you need to update the [test data](https://github.com/ChainSafe/lodestar/blob/unstable/packages/api/test/unit/beacon/testData/debug.ts)

acolytec3 replied:
Switched this over to be part of the generic `testData`

acolytec3 replied:
Do we need a more comprehensive test that gives a populated response?  I'm not sure the best way to approach this given how your tests are structured but happy to add if it's fruitful and there's an obvious place for it.

nflaig replied:
> Do we need a more comprehensive test that gives a populated response?

The most comprehensive would be to add an e2e test, eg. see https://github.com/ChainSafe/lodestar/blob/64eb0153dd97b7525cb61625b102a6a521869c05/packages/beacon-node/test/e2e/api/impl/beacon/state/endpoint.test.ts#L46

I do think those would be nice-to-have but there is not much internal logic as it's mostly just wiring up existing functions which are tested separately arleady.

Imo for this debug and v0 endpoint we should be fine with the generic tests but if you wanna add an e2e test go for it

nflaig commented on packages/api/src/beacon/routes/debug.ts:
```diff
@@ -135,6 +141,13 @@ export type Endpoints = {
+    "GET",
+    StateArgs,
+    {params: {state_id: string}},
+    ValueOf<typeof HistoricalSummariesResponseType>,
```
to align with other apis, best to declare type separate (see above)

acolytec3 replied:
Done

nflaig commented on packages/beacon-node/src/api/impl/debug/index.ts:
```diff
@@ -85,5 +86,28 @@ export function getDebugApi({
+        throw new Error("Historical summaries are not supported before Capella");
+      }
+      const fork = config.getForkName(slot) as Exclude<ForkName, "phase0" | "altair" | "bellatrix">;
+      const stateView = state instanceof Uint8Array ? ssz[fork].BeaconState.deserializeToViewDU(state) : state;
```
This needs to follow similar pattern as here
https://github.com/ChainSafe/lodestar/blob/64eb0153dd97b7525cb61625b102a6a521869c05/packages/beacon-node/src/api/impl/validator/index.ts#L1021-L1024

otherwise we it will be pretty expensive memory wise

acolytec3 replied:
Fixed

nflaig reviewed (commented):
Mostly looks good to me, just few remarks regarding loading the state from bytes and need to make CI happy

acolytec3 committed:
Address feedback

acolytec3 commented:
@nflaig general question, should I go ahead and move over to the `lodestar` namespace for now, following what `nimbus` is doing?  I don't have an opinion on which namespace (nor do I know if this proposal would ultimately end up under `debug` or under maybe `beacon` as a light client API (since presumably this will be used mainly by Portal Network clients).  

One further question, does it make sense to add the `meta` to the response mirroring the `getState` API?  We don't use this directly in Portal Network so I don't have the context to know if this is a useful addition to the overal API.

nflaig commented:
> general question, should I go ahead and move over to the lodestar namespace for now, following what nimbus is doing?

Yeah, might be better to go with lodestar namespace if it's anyhow not standardized between clients.

> One further question, does it make sense to add the meta to the response mirroring the getState API? We don't use this directly in Portal Network so I don't have the context to know if this is a useful addition to the overal API.

If it would become standardized as part of beacon-api then it will for sure have the meta as well, if you don't need it we don't have to add it now, we can also always add it later on if there is demand.

acolytec3 committed:
Move getHistoricalSummaries to lodestar namespace

acolytec3 committed:
add lodestar namespace unit test

acolytec3 commented:
> > general question, should I go ahead and move over to the lodestar namespace for now, following what nimbus is doing?
> 
> Yeah, might be better to go with lodestar namespace if it's anyhow not standardized between clients.
> 
> > One further question, does it make sense to add the meta to the response mirroring the getState API? We don't use this directly in Portal Network so I don't have the context to know if this is a useful addition to the overal API.
> 
> If it would become standardized as part of beacon-api then it will for sure have the meta as well, if you don't need it we don't have to add it now, we can also always add it later on if there is demand.

Sounds good.  I have moved everything to lodestar and added a new unit test file for `lodestar` since it doesn't appear there was a unit test at all for that namespace.  I opted to just create a specific test for that endpoint since running the `GenericServerTest` looks like it would require me to add unit testData for all of the `lodestar` endpoints which feels like it's beyond the scope of this PR.

acolytec3 committed:
update route name to lodestar namespace

wemeetagain commented on packages/beacon-node/src/api/impl/lodestar/index.ts:
```diff
@@ -187,6 +190,33 @@ export function getLodestarApi({
+      const fork = config.getForkName(slot) as Exclude<ForkName, "phase0" | "altair" | "bellatrix">;
+
+      const stateView =
+        state instanceof Uint8Array ? loadState(config, chain.getHeadState(), state).state : state.clone();
```
```suggestion
        (state instanceof Uint8Array ? loadState(config, chain.getHeadState(), state).state : state.clone()) as BeaconStateCapella;
```

You can avoid the `any` below by casting this to a `BeaconStateCapella` here

acolytec3 replied:
It seems like the proof size for `HistoricalSummaries` changed in `deneb` (and will maybe again in Electra).  Is it still safe to cast it to `BeaconStateCapella` here since the shape of the state is changing somewhat?

acolytec3 committed:
cast state object as Capella state

nflaig committed:
Merge branch 'unstable' into historicalSummaries-endpoint

nflaig committed:
Lint

nflaig committed:
json properties need to be lower case

nflaig committed:
Make it v1 since it's now part of lodestar namespace

nflaig committed:
Group with other /lodestar endpoints

nflaig committed:
Simplify beacon node impl

nflaig committed:
Rename return type

nflaig committed:
Update test description

nflaig committed:
Fix variable name

nflaig commented on packages/api/src/beacon/routes/lodestar.ts:
```diff
@@ -75,6 +79,16 @@ export type LodestarNodePeer = NodePeer & {
 
+const HistoricalSummariesResponseType = new ContainerType(
+  {
+    historicalSummaries: ssz.capella.HistoricalSummaries,
```
@acolytec3 I converted this to lower case to align it with existing json responses

nflaig commented on packages/api/src/beacon/routes/lodestar.ts:
```diff
@@ -365,6 +389,21 @@ export function getDefinitions(_config: ChainForkConfig): RouteDefinitions<Endpo
       resp: JsonOnlyResponseCodec,
     },
+    getHistoricalSummaries: {
+      url: "/eth/v1/lodestar/historical_summaries/{state_id}",
```
Changed to use v1 since it's now part of `lodestar` namespace

nflaig reviewed (approved):
LGTM

acolytec3 commented:
Thanks for the approval!  @nflaig once this gets merged, any idea how long before the public lodestar bootnodes will make it available?

nflaig commented:
> any idea how long before the public lodestar bootnodes will make it available?

We plan to cut a rc at the end of the week, we discussed in the standup today to get your PR in for that as well, so the next release if testing goes well should be some time next week which means the public nodes would be updated next week as well.

nflaig committed:
Merge branch 'unstable' into historicalSummaries-endpoint

wemeetagain commented:
:tada: This PR is included in [v1.24.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.24.0) :tada:

acolytec3 commented:
Thank you sir :vulcan_salute:
